[DEBUG]Support a debug framework for TVM Runtime

siju-samuel opened the pull request:
[RFC Issue]( https://github.com/dmlc/tvm/issues/1315)
#https://github.com/dmlc/tvm/issues/1315

A debug framework for TVM Runtime.
The current version support the following:
- [x]   Show fused graph summary in curses UI
- [x]   Perform debug run and show node details including inputs & outputs tensors
- [x]   Provide flexibility to run without debug option

**Design**
1. `debug_runtime.py ` : This is the extension of `graph_runtime.py`. If user enable the debug, then all the graph_runtime interfaces are controlled from here and  the data dumping will be maintained from here. The data dumping to different UX, whether curses or tensorboard is maintained from this file.

2. `graph_runtime_debug.cc` : This is the extension of graph_runtime.cc file. To compile this file, need to change the make file USE_GRAPH_RUNTIME_DEBUG to ON

3. `curses` :  This folder contains the curses framework. curses will do the data parsing and visualisation for various features like analyser, stepper & profiler.

The current implementation takes care of the following.
- [x] No memory impact for normal run-time since the existing data structures are kept intact.
- [x] No lib size impact for normal run-time, as all the debug modificatios are protected with compilation macro USE_GRAPH_RUNTIME_DEBUG.
- [x] No execution time impact for normal run-time  as there are no change in the existing functions.
- [x] Test case added for debug module creation and partial validation (run cant be tested since the curses ui will come up block the execution)
- [x] The data logging and UX work independently, Output data logged into temporary folder and UX read this data and proceed visualisation.
- [x] Tutorial on how to use.

Currently this PR supports only graph visualisation using curses UX, later will add support for profiling, stepping and tensorboard support.

tqchen commented:
@merrymercy @Huyuwei @yzhliu can you please review this?

yzhliu commented on docs/dev/tvmdbg.md:
```diff
@@ -0,0 +1,57 @@
+
+https://github.com/tensorflow/tensorflow
+https://github.com/tensorflow/tensorboard
+https://github.com/awslabs/mxboard
```
break to three lines.

yzhliu commented on docs/dev/tvmdbg.md:
````diff
@@ -0,0 +1,57 @@
+2. Make tvm so that it will make the `libtvm_runtime.so`
+
+3. In the graph build file instead of `from tvm.contrib import graph_runtime` import the debug_runtime `from tvm.contrib.debugging import debug_runtime as graph_runtime`
+```
````
use ```python to highlight the py keywords.

yzhliu commented on docs/dev/tvmdbg.md:
```diff
@@ -0,0 +1,57 @@
+The above two modifications will bring up the debug UI during run.
+
+The HOME page of tvmdbg looks like below.
+ ![](https://raw.githubusercontent.com/siju-samuel/tvmdbg/master/docs/dev/_images/tvm_dbg1.png)
```
@tqchen do we have some 'official' place to store these images?

srkreddy1238 replied:
dmlc/web-data can be used for all content.

tqchen replied:
Let us use https://github.com/tvmai/tvmai.github.io for this

yzhliu commented on python/tvm/contrib/debugger/curses/ui/analyzer/analyzer.py:
```diff
@@ -0,0 +1,1346 @@
+    #menu.append(
+    #    ui_common.MenuItem("home", "home"))
+    #menu.append(
+    #    ui_common.MenuItem("help", "help"))
```
remove unused lines?

yzhliu commented on python/tvm/contrib/debugger/curses/ui/command_parser.py:
```diff
@@ -0,0 +1,568 @@
+      ValueError: on failure to parse the input `size_str`.
+    """
+
+    size_str = size_str.strip()
```
better to normalize by ```size_str.upper()``` and check the upper case K, B, M, G, etc.

yzhliu commented on python/tvm/contrib/debugger/debug_runtime.py:
```diff
@@ -0,0 +1,541 @@
+        """
+        dbg_out_buffer_list = []
+        for i in range(len(shapes_list[1])):
+            dbg_out_buffer_list.append(nd.empty(shapes_list[1][i], dltype_list[1][i]))
```
just to make sure I understand correctly: these buffers always stay on cpu, which are for storing the network outputs for debug.

siju-samuel replied:
yes. always stay in cpu. for storing the outputs of each fused op after execution, 
This will be updated after each op execution.

tqchen commented on python/tvm/contrib/debugger/curses/ui/tensor_data.py:
```diff
@@ -0,0 +1,555 @@
+
+    Parameters
+    ----------
+      tensor: The tensor to be displayed, as a numpy ndarray or other
```
parameter format https://docs.tvm.ai/contribute/document.html#document-python

tqchen commented:
Thanks, @siju-samuel for the contribution and sorry for the delayed review. By looking into the current PR I think we can split it into two logical parts.

The current workflow of debugger works by running debug-runtime collect all the data we need, and then invoke a visualizer to step through the collected result.

So logically, we can draw a picture like

```tvm.contrib.debugger.DebugGraphRuntime -> tvm.contrib.debugger.DebugResult(currently GraphModuleDebugDumpDatum) -> UX```

Let us change this PR to only contain ```tvm.contrib.debugger.DebugGraphRuntime -> tvm.contrib.debugger.DebugResult(currently GraphModuleDebugDumpDatum) ``` and document the debug result exchange format clearly.

We can bring this in first. Once this PR get merged. The users can readily try out the debugger feature if we wrap the curse visualizer as a separate python package that you maintain. I would recommend we collect feedbacks from users and improves the curses UX, before we stablize and merge it back to master.

The tensorboard support can also be done in concurrent PRs. I recommend this path because this will force us think carefully about this separation of UX and data logging, and also ensures the high quality components, as @yzhliu and I can review the data logging components but less familar with the UX part

tqchen commented:
@srkreddy1238 @PariksheetPinjari909 can you also please help review?

siju-samuel commented:
@tqchen the review comments are addressed and all the curses functions are made according to tvm guidelines.
debug_runtime.py & debug_result.py handles the interfaces with graph_runtime.

> The users can readily try out the debugger feature if we wrap the curse visualizer as a separate python package that you maintain. I would recommend we collect feedbacks from users and improves the curses UX, before we stablize and merge it back to master.

If we keep as a separate package, users won't use this feature extensively, so I suggest we can keep it as part of tvm package int a separate folder , the complete maintenance responsibility of tvmdbg, I will handle. This have no impact to the actual runtime as this is disabled by default. I want users to use this and give their feedback so that it could be improved further.

tqchen commented:
@siju-samuel I understand your concerns in here.  Let us aim to merge this in, but still do it in two step PR, (raise the UX part in a separate PR).

The first PR should contain
- [ ] ```tvm.contrib.debugger.DebugGraphRuntime -> tvm.contrib.debugger.DebugResult(currently GraphModuleDebugDumpDatum)```
- [ ] Document on the debugger exchange format, on how a visualizer can read these in and visualize the result
- [ ] Testcases on verifying the exchange format logging.

The reason to push for this is that to force us to make clear API separation and pave the path for tensorboard integration. I think it would also help a lot to get reviewers to review the UX part once the exchange format is clear

siju-samuel commented:
@tqchen  @yzhliu 
The first PR is updated. please review and give your opinion. 
- [x]   tvm.contrib.debugger.DebugGraphRuntime -> tvm.contrib.debugger.DebugResult
- [x]   Documentation on the debugger exchange format
- [x]   Testcases on verifying the data dumps is added

The second PR about ncurses will be done based on the review comments received for this one.
TIA

tqchen commented:
@siju-samuel can you remove the usage in tvmdbg.md for now(as they are UI related) and instead add a section about Debug Exchange Format, and document all the exchange file format(how each field of json correspond to the which each info)

siju-samuel commented:
@tqchen updated as per your comment.

tqchen commented:
Thanks for the update, we still need to improve the specification to give all the details, instead of examples. Please refer to https://docs.tvm.ai/dev/nnvm_json_spec.html 

As for the tensor storage format, one way we might be able to do is to provide a base64 serialization format of the tensor, so that can be embedded into json, would that be helpful?

See related PR here https://github.com/dmlc/tvm/pull/1452

siju-samuel commented:
> Thanks for the update, we still need to improve the specification to give all the details, instead of examples. Please refer to https://docs.tvm.ai/dev/nnvm_json_spec.html
Thanks for the review. I updated the json format used in exchange. Please suggest if i need to update further. 

> As for the tensor storage format, one way we might be able to do is to provide a base64 serialization format of the tensor, so that can be embedded into json, would that be helpful?
if we add the storage in tensor along with the graph dump, the file will become so big for big networks. When i checked tensorflow/mxboard they are also creating separate files for tensors. Another thing is graph wont change. but tensors can keep changing based on input data. So if user run graph again and again with different inputs, we need to update the complete graph dump file. numpy dumping format is easy and wont consume much memory. tensorboard, curses can directly load the numpy format without any conversion.

tqchen commented on docs/dev/tvmdbg.md:
```diff
@@ -0,0 +1,105 @@
+### TVMDBG
```
change to level one title # Debugger

tqchen replied:
Maybe I am asking too much, but is it possible to change the doc to reStucturedText? It will make things easier to refer to specific section of the document

tqchen commented on docs/dev/tvmdbg.md:
```diff
@@ -0,0 +1,105 @@
+
+TVM Debugger (TVMDBG) is an interface for debugging TVM's computation graph execution. It helps to provide access to graph structures and tensor values at the TVM runtime.
+
+### Debug Exchange Format
```
```## Exchange Format```

tqchen commented on docs/dev/tvmdbg.md:
```diff
@@ -0,0 +1,105 @@
+TVM Debugger (TVMDBG) is an interface for debugging TVM's computation graph execution. It helps to provide access to graph structures and tensor values at the TVM runtime.
+
+### Debug Exchange Format
+**1. Graph information**
```
```### Computational Graph ```

tqchen commented on docs/dev/tvmdbg.md:
````diff
@@ -0,0 +1,105 @@
+
+```
+
+**2. Tensor dumping**
````
```### Tensor Data````

tqchen commented on docs/dev/tvmdbg.md:
```diff
@@ -0,0 +1,105 @@
+Each node in the graph will be dumped to individual files, in the dump folder. These files will be created after execution of each node.
+
+
+### How to use TVMDBG?
```
`## Enable Debugger`

tqchen commented on docs/dev/tvmdbg.md:
```diff
@@ -0,0 +1,105 @@
+More about numpy format can be read in the below link.
+[numpy-format](https://docs.scipy.org/doc/numpy/neps/npy-format.html)
+Each node in the graph will be dumped to individual files, in the dump folder. These files will be created after execution of each node.
+
```
Add a section on how to get these data from debugger interface. This is useful for debugger UX developers

tqchen commented on python/tvm/contrib/debugger/debug_result.py:
```diff
@@ -0,0 +1,147 @@
+
+GRAPH_DUMP_FILE_NAME = '_tvmdbg_graph_dump.json'
+
+class DebugResult():
```
class DebugResult(object):

tqchen commented on python/tvm/contrib/debugger/debug_runtime.py:
```diff
@@ -0,0 +1,424 @@
+        This function will get called before run performs.
+        GraphRuntime copy the execution out to the allocated memory for each nodes.
+
+        Parameters
```
if there is no parameters, do not include parameters section

tqchen commented on python/tvm/contrib/debugger/debug_runtime.py:
```diff
@@ -0,0 +1,424 @@
+        Parameters
+        ----------
+
+        cli_obj : obj
```
parameters do not match the signature

tqchen commented on python/tvm/contrib/debugger/debug_runtime.py:
```diff
@@ -0,0 +1,424 @@
+        """Exits the dump folder and all its contents"""
+        self._remove_dump_root()
+
+class DebugGraphUXWrapper(object):
```
Remove this class for now as it is unclear how this class can be used, alternatively, give an example on how to make simple UX out of the class

tqchen commented:
@yzhliu @srkreddy1238 please take another look of this PR

yzhliu commented on docs/dev/tvmdbg.md:
```diff
@@ -0,0 +1,105 @@
+The Graph JSON format is explained below 
+1. nodes
+  Nodes are either placeholders or computational nodes in NNVM graph. The nodes are stored as a list. A node contains the below information
+   `op` - operation type, `null` means its a placeholder/variable/input node and `tvm_op` means this node can be executed
```
its -> it is

yzhliu commented on python/tvm/contrib/debugger/debug_result.py:
```diff
@@ -0,0 +1,147 @@
+        """
+        return self._shapes_list
+
+    def get_graph_node_dltypes(self):
```
feel it is better to name it 'dtypes', keep the name convention the same across the project.

yzhliu commented on python/tvm/contrib/debugger/debug_result.py:
```diff
@@ -0,0 +1,147 @@
+                eid += 1
+                order += time
+                key = node['name'] + "_" + str(j) + "__" + str(order) + ".npy"
+                dump_file = str(self._dump_path + key.replace("/", "_"))
```
use `os.sep` instead of `/` ?

tqchen commented:
For storing tensor data, can we reuse https://docs.tvm.ai/api/python/nnvm/compiler.html?highlight=save_param_dict#nnvm.compiler.save_param_dict and load_param_dict, this way we can avoid having to store too many files

siju-samuel commented:
@tqchen the output tensors, i changed to `save_param_dict` from numpy.  You are right, too many files is  overhead. Review comments are updated.   
but the CI failed because i have to `import nnvm` for `save_param_dict`, one solution is i will write a python wrapper in tvm and use tvm::runtime::SaveDLTensor and serialize?

siju-samuel commented:
@tqchen I have updated the code  based on review comemnts.
`save_param_dict` needs `import nnvm` so im not able to use it in tvm
One method is all numpy files can be appended together.

tqchen commented:
@siju-samuel We can bring the save_param_dict into tvm to fix the problem, it is good to have a consistent data storage that is not python dependent

siju-samuel commented:
@tqchen tensors are saved using save_param_dict and the code/documentation is updated.

tqchen commented:
@yzhliu @srkreddy1238 can you do another round of review?

srkreddy1238 commented on python/tvm/contrib/debugger/debug_result.py:
```diff
@@ -0,0 +1,160 @@
+    ----------
+    graph_json : str
+        The graph to be deployed in json format output by nnvm graph. The graph can only contain
+        one operator(tvm_op) that points to the name of PackedFunc in the libmod.
```
Readability: "one operator" confuse a bit. Rephrase it similar to below suggestion.

"The graph can only contain one operator(tvm_op) that points to the name of PackedFunc in the libmod."
=>
"Each operator (tvm_op) in the graph will have a one to one mapping with the symbol in libmod which is used to construct a "PackedFunc" "

srkreddy1238 commented on python/tvm/contrib/debugger/debug_result.py:
```diff
@@ -0,0 +1,160 @@
+                node['op'] = "param"
+            else:
+                node['op'] = node['attrs']['func_name']
+            node['name'] = node['name'].replace("/", "_")
```
Pls use unicodedata API to normalize.

srkreddy1238 commented on python/tvm/contrib/debugger/debug_result.py:
```diff
@@ -0,0 +1,160 @@
+                            else int(node['attrs']['num_outputs'])
+            for j in range(num_outputs):
+                order += time
+                key = node['name'].replace("/", "_") + "_" + str(j) + "__" + str(order)
```
I see the normalization for node name is done in _update_graph_json above.
If not pls use unicodedata API here too.

srkreddy1238 commented on python/tvm/contrib/debugger/debug_result.py:
```diff
@@ -0,0 +1,160 @@
+            json.dump(graph, outfile, indent=2, sort_keys=False)
+
+    def _cleanup_tensors(self):
+        """Remove the tensor dumps files(grah wont be removed)
```
dumps -> dump
grah -> graph

srkreddy1238 commented on python/tvm/contrib/debugger/debug_runtime.py:
```diff
@@ -0,0 +1,327 @@
+        return str(ctx).upper().replace("(", ":").replace(")", "")
+
+    def _prepare_data_and_ui(self, graph_json, ctx, dbg_ux):
+        """Create the framework for debug data dumpling and initialize the frontend
```
dumpling -> dumping

srkreddy1238 commented on python/tvm/contrib/debugger/debug_runtime.py:
```diff
@@ -0,0 +1,327 @@
+    graph_json_str : str or graph class
+        The graph to be deployed in json format output by nnvm graph.
+        The graph can only contain one operator(tvm_op) that
+        points to the name of PackedFunc in the libmod.
```
graph json definition is detailed above already (tvmdbg.rst), just keep it simple.

"Content of graph json file in string format" or similar.

srkreddy1238 commented on python/tvm/contrib/debugger/debug_runtime.py:
```diff
@@ -0,0 +1,327 @@
+        Parameters
+        ----------
+        graph_json : str or graph class
+            The graph to be deployed in json format output by nnvm graph.
```
Same as above. Simplify the description.

srkreddy1238 commented on python/tvm/contrib/debugger/debug_runtime.py:
```diff
@@ -0,0 +1,327 @@
+            File path to create
+        """
+        if not os.path.exists(directory):
+            os.makedirs(directory)
```
Pls fix the "mode" for the directory.

srkreddy1238 commented on python/tvm/contrib/debugger/debug_runtime.py:
```diff
@@ -0,0 +1,327 @@
+            os.makedirs(directory)
+
+    def _get_dump_path(self, ctx):
+        """Dump json formatted graph.
```
Correct the description.

srkreddy1238 commented on python/tvm/contrib/debugger/debug_runtime.py:
```diff
@@ -0,0 +1,327 @@
+        self._run_debug()
+
+    def set_input(self, key=None, value=None, **params):
+        """Set inputs to the module via kwargs
```
kwargs ??

siju-samuel commented:
@srkreddy1238 thanks for the review. I have updated based on the review comments. Please have a look.

siju-samuel commented:
Thanks for the support. Its re-based and pushed. 
@tqchen  @srkreddy1238 @yzhliu  Could you kindly check the updated code. Thanks

siju-samuel commented:
@tqchen @yzhliu Could you please do another round of review and merge this PR?

tqchen commented:
Sorry for the delayed review. Debugging/profiling is important and we want to get things right:) 

- After #1610, we should be able to directly support get_ndarray from the debugger, this might involves refactor the graph runtime a bit to directly use ndarray, this removes the need of adoc set_debug_tensor(Note that it is not safe to pass DLTensor(should use NDArray instead)) to retain ownership. cc @srkreddy1238 
- Use chrono to do the timing, see https://github.com/dmlc/tvm/blob/master/src/runtime/rpc/rpc_session.cc#L1200
   - Always synchronize after each step run, to get the time measurement right(note that we need to sync to get GPUs).
-  Let us remove most of the Ux objects for now, create a simple demo to run the debugger and print out the per layer timing.
   - Try to ,minimize API change over graph runtime.
   - This seems to be the feature most people want

merrymercy commented on src/runtime/graph/debug/graph_runtime_debug.cc:
```diff
@@ -0,0 +1,126 @@
+    if (op_execs()[index]) {
+      op_execs()[index]();
+    }
+    start_time = int64_t(tp.tv_sec * 1000000L + tp.tv_usec);
```
This line should be placed before `op_execs`

merrymercy commented on docs/dev/tvmdbg.rst:
```diff
@@ -0,0 +1,133 @@
+3. In frontend script file instead of
+   ``from tvm.contrib import graph_runtime`` import the
+   ``debug_runtime``
+   ``from tvm.contrib.debugging import debug_runtime as graph_runtime``
```
debugging -> debugger

merrymercy commented on python/tvm/contrib/debugger/debug_runtime.py:
```diff
@@ -0,0 +1,323 @@
+        super(GraphModuleDebug, self).set_input(key, value, **params)
+
+        if key:
+            self.ui_obj.set_input(key, value)
```
`self.oi_obj` can be none

siju-samuel replied:
@merrymercy Currently only the graph and output of each layer is saved to file. The UI object is none. This two lines, I will remove. 
I will update this whole PR based on tqchen's comment.

siju-samuel commented:
@tqchen Thanks for the update. Please have a look again, I updated the code based on your comment.

> After #1610, we should be able to directly support get_ndarray from the debugger, this might involves refactor the graph runtime a bit to directly use ndarray, this removes the need of adoc set_debug_tensor(Note that it is not safe to pass DLTensor(should use NDArray instead)) to retain ownership. cc 

I couldn't find any function in ndarray.cc which will make NDArray from DLTensor, so I updated ndarray.cc  to added a method for this.
set_debug_tensor is completely removed and NDArray is used.

> Use chrono to do the timing, see https://github.com/dmlc/tvm/blob/master/src/runtime/rpc/rpc_session.cc#L1200

Updated with chrono for timing

> Always synchronize after each step run, to get the time measurement right(note that we need to sync to get GPUs).

Syncing is taken care

> Let us remove most of the Ux objects, for now, create a simple demo to run the debugger and print out the per layer timing.

Ux objects are removed, some interfaces, its kept as argument(dbg_ux): . This debugger was written for supporting ncurses ui and tensorboard, so the ux objects were there. Now this does the dumping of nodes, and display. 
Now it will show some simple display about the time consumed for each layer, to visualize the relation of each node and tensor values, still we need to use curses/tensorboard.
The current output format is below.
```
Node Name        Ops                                Time(us)             Shape               Inputs       Outputs      
---------        ---                                --------             -----               ------       -------      
relu0            fuse_conv2d_broadcast_add_relu     14840.649            (1, 1, 224, 224)    3            1            
relu1            fuse_conv2d_broadcast_add_relu_1   224299.577           (64, 1, 5, 5)       3            1            
relu2            fuse_conv2d_broadcast_add_relu_2   111466.68699999999   (64, 1, 1)          3            1            
broadcast_add3   fuse_conv2d_broadcast_add          25375.287            (1, 64, 224, 224)   3            1            
reshape1         fuse_reshape_transpose_reshape     1309.9260000000002   (64, 64, 3, 3)      1            1     
````

> Try to minimize API change over graph runtime.

There is no modification in the logic of graph_runtime.cc or graph_runtime.h, To extend the graph_runtime, i moved the definitions from .cc to .h, But this change looks big. 

> After reading the things, I feel we do not necessarily need to serialize the graph and ndarray and can just provide everything in an in-memory format.

I kept serializing and dumping as of now, since we can extend this framework to use in tensorboard/curses.

tqchen commented:
Comments wrt to NDArray, we might need to send a PR first to change graph runtime to use NDArray instead of DLTensor* in allocation, note that NDArray can be converted to DLTensor. This is similar to @srkreddy1238 's discussion on getting shape API so maybe you can work together to bring a PR.

I like the DebugResult object and the simple print function that uses to print the result.  In terms of **separation of concerns**.
- It seems that the process is really graph_runtime_debug.run()->DebugResult
- DebugResult -> ux-viewer

If that is the case, we should not couple the UX code in the debug run and focus on making the DebugRuntime containing a single interface that get you the DebugResult, and then followup with viewers that takes DebugResult and do the visualization.

siju-samuel commented:
> Comments wrt to NDArray, we might need to send a PR first to change graph runtime to use NDArray instead of DLTensor* in allocation, note that NDArray can be converted to DLTensor.

Will work together with @srkreddy1238 and first will bring the PR for changing from DLTensor to NDArray in runtime. If the ndarray is returned from runtime, the shape query api can be eliminated.

> If that is the case, we should not couple the UX code in the debug run and focus on making the DebugRuntime containing a single interface that get you the DebugResult, and then followup with viewers that takes DebugResult and do the visualization.

All the UX code is now removed. Now debug_runtime.py is the extension of graph_runtime.py and DebugResults will dump and display the outputs. With this design UX can be completely independent and no dependency with debug_runtime.

siju-samuel commented:
@tqchen  @srkreddy1238 @merrymercy could you please do another round of review on latest code and conclude this PR?

merrymercy commented on python/tvm/contrib/debugger/debug_result.py:
```diff
@@ -0,0 +1,205 @@
+                    continue
+                name = node['name']
+                shape = str(self._output_tensor_list[eid].shape)
+                time_us = time * 1000000
```
`time_us = round(time * 1000000, 2)`

merrymercy commented on src/runtime/graph/debug/graph_runtime_debug.cc:
```diff
@@ -0,0 +1,133 @@
+    auto tbegin = std::chrono::high_resolution_clock::now();
+    if (op_execs()[index]) {
+      op_execs()[index]();
+    }
```
For gpu, we should sync stream by calling this function on ctx
https://github.com/dmlc/tvm/blob/a03c60ba1c83d39b857b827f144bf1b371d0b0c7/python/tvm/_ffi/runtime_ctypes.py#L200

siju-samuel replied:
@merrymercy  I invoked sync from python. I got the issue with that. The timestamp won't reflect if I invoked from python. I will update. Thanks.

merrymercy commented on python/tvm/contrib/debugger/debug_result.py:
```diff
@@ -0,0 +1,205 @@
+    def display_debug_result(self):
+        """Displays the debugger result"
+        """
+        header = ["Node Name", "Ops", "Time(us)", "Shape", "Inputs", "Outputs"]
```
Add a column to show the percentage "Time (%s")

merrymercy commented on python/tvm/contrib/debugger/debug_result.py:
```diff
@@ -0,0 +1,205 @@
+                eid += 1
+        fmt = ""
+        for i, _ in enumerate(header):
+            max_len = 10
```
max_len = len(header[i])

siju-samuel commented:
@merrymercy Thanks for the review and i have updated the code based on your comments.
@srkreddy1238 @tqchen Please review once again.

tqchen commented on python/tvm/contrib/debugger/debug_runtime.py:
```diff
@@ -0,0 +1,188 @@
+
+    func_obj = fcreate(graph_json_str, libmod, *device_type_id)
+    return GraphModuleDebug(func_obj, ctx, graph_json_str, dump_root)
+
```
nit: two lines between modules

tqchen commented on python/tvm/contrib/debugger/debug_result.py:
```diff
@@ -0,0 +1,207 @@
+        for row in data:
+            print(fmt.format(*row))
+
+def save_tensors(params):
```
move the PackedFunc to part of src/api/api_base.cc

tqchen commented on python/tvm/contrib/debugger/debug_result.py:
```diff
@@ -0,0 +1,207 @@
+        """
+        self._output_tensor_list.append(out_tensor)
+
+    def set_time(self, time):
```
this seems to be too simple function to add a level of indirection, directly inline at call site

zhiics replied:
+1

tqchen commented on src/runtime/graph/debug/graph_runtime_debug.cc:
```diff
@@ -0,0 +1,135 @@
+        << args.num_args;
+    *rv = GraphRuntimeDebugCreate(args[0], args[1], GetAllContext(args));
+  });
+TVM_REGISTER_GLOBAL("tvm.graph_runtime_debug._save_param_dict")
```
move this function to src/api/api_base.cc for now

tqchen commented on src/runtime/graph/graph_runtime.h:
```diff
@@ -24,6 +39,374 @@ struct TVMOpParam {
+  /*!
+   * \brief Copy index-th node to data_out.
+   *
+   * This method will do a partial run of the the graph
```
move debug signature to GraphRuntimeDebug

tqchen commented:
Thanks @siju-samuel for being patient during the review process and keep improving the PR, I have made some followup comments. @zhiics @merrymercy @srkreddy1238 please take another round of look and let us aim to merge this in soon

zhiics commented on python/tvm/contrib/debugger/debug_result.py:
```diff
@@ -0,0 +1,207 @@
+    """Graph debug data module.
+
+    Data dump module manage all the debug data formatting.
+    Output data and input graphs are formatted and the dump to files.
```
s/the dump/dumped

zhiics commented on docs/dev/debugger.rst:
```diff
@@ -0,0 +1,149 @@
+- ``dtype`` - Datatype of each node (enum value)
+- ``dltype`` - Datatype of each node in order
+- ``shape`` - Shape of each node k order
+
```
Please add:
``device_index`` - device assignment for each entry in the graph.

zhiics commented on python/tvm/contrib/debugger/debug_result.py:
```diff
@@ -0,0 +1,207 @@
+        self._time_list.append(time)
+
+    def dump_output_tensor(self):
+        """Dump the outputs to a temporary folder, the tensors is in numpy format
```
s/is/are

zhiics commented on python/tvm/contrib/debugger/debug_runtime.py:
```diff
@@ -0,0 +1,188 @@
+        ctx : TVMContext
+            The context this module is under.
+        """
+        #make the dump folder if not given
```
# make

zhiics commented on python/tvm/contrib/debugger/debug_runtime.py:
```diff
@@ -0,0 +1,188 @@
+        if not self._dump_root:
+            self._dump_root = tempfile.mktemp(prefix=_DUMP_ROOT_PREFIX)
+
+        #format the context
```
ditto

zhiics commented on python/tvm/contrib/debugger/debug_runtime.py:
```diff
@@ -0,0 +1,188 @@
+        #format the context
+        ctx = self._format_context(ctx)
+
+        #updates the dumping directories
```
ditto

zhiics commented on python/tvm/contrib/graph_runtime.py:
```diff
@@ -31,6 +31,31 @@ def create(graph_json_str, libmod, ctx):
+    -------
+    ctx : list of TVMContext
+    num_rpc_ctx : Number of rpc contexts
+    device_type_id : List of device type
```
List of device type and id?

zhiics commented on src/runtime/graph/graph_runtime.h:
```diff
@@ -24,6 +39,374 @@ struct TVMOpParam {
+   *
+   * \return NDArray corresponding to given input node index.
+   */
+  NDArray GetInput(int index);
```
GetInput(int index) const?

zhiics commented on src/runtime/graph/graph_runtime.h:
```diff
@@ -24,6 +39,374 @@ struct TVMOpParam {
+   *
+   * \return NDArray corresponding to given output node index.
+   */
+  NDArray GetOutput(int index);
```
const?

zhiics commented on src/runtime/graph/graph_runtime.h:
```diff
@@ -24,6 +39,374 @@ struct TVMOpParam {
+   * \param name The name of the node.
+   * \return The index of node.
+   */
+  int GetNodeIndex(const std::string& name);
```
const?

zhiics commented:
@siju-samuel Thanks for the nice work. I only have some nit comments.

zhiics commented:
@siju-samuel @tqchen Out of curiosity, is it necessary to also show the start and end timestamp of each operator so that we could be able to see the "idle time" in between (e.g. this kind of data could be exported to some UI tools conveniently)? This might be also helpful for debugging.

srkreddy1238 commented:
I had another look with new changes. @zhiics already covered the heterogeneous related changes.

We could add another test case (may be later) to demonstrate the debug tensors over heterogeneous with intermediate node on non-cpu context.

siju-samuel commented:
@tqchen @zhiics @srkreddy1238 @merrymercy @yzhliu Thanks for the reviews.
I have updated the code based on review comments. Please check once again. Thanks.

zhiics commented:
@siju-samuel LGTM.

tqchen commented:
Thanks @siju-samuel for being patient over the review process, thanks @zhiics @merrymercy @srkreddy1238 @yzhliu for the reviews. this is  now merged
